common/flatpak-prune.c

```diff
@@ -0,0 +1,813 @@
+      *out_exists = FALSE;
+      return TRUE;
+    }
+  if (!glnx_dirfd_iterator_init_take_fd (&fd, dfd_iter, error))
+    return FALSE;
+  *out_exists = TRUE;
+  return TRUE;
+}
+
+/* Object name helpers */
+
+static guint
+_ostree_object_name_hash (gconstpointer a)
+{
+  VarObjectNameRef ref = var_object_name_from_gvariant ((GVariant *)a);
+
+  return g_str_hash (var_object_name_get_checksum (ref)) + (guint)var_object_name_get_objtype (ref);
+}
+
+static gboolean
+_ostree_object_name_equal (gconstpointer a,
+                           gconstpointer b)
+{
+  VarObjectNameRef ref_a = var_object_name_from_gvariant ((GVariant *)a);
+  VarObjectNameRef ref_b = var_object_name_from_gvariant ((GVariant *)a);
+
+  return
+    g_str_equal (var_object_name_get_checksum (ref_a), var_object_name_get_checksum (ref_b)) &&
+    var_object_name_get_objtype (ref_a) == var_object_name_get_objtype (ref_b);
+}
+
+static GHashTable *
+reachable_commits_new (void)
+{
+  return g_hash_table_new_full (_ostree_object_name_hash, _ostree_object_name_equal,
+                                NULL, (GDestroyNotify)g_variant_unref);
+}
+
+/* Wrapper to handle flock vs OFD locking based on GLnxLockFile */
+static gboolean
+do_repo_lock (int fd,
+              int flags)
+{
+  int res;
+
+#ifdef F_OFD_SETLK
+  struct flock fl = {
+    .l_type = (flags & ~LOCK_NB) == LOCK_EX ? F_WRLCK : F_RDLCK,
+    .l_whence = SEEK_SET,
+    .l_start = 0,
+    .l_len = 0,
+  };
+
+  res = TEMP_FAILURE_RETRY (fcntl (fd, (flags & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW, &fl));
+#else
+  res = -1;
+  errno = EINVAL;
+#endif
+
+  /* Fallback to flock when OFD locks not available */
+  if (res < 0)
+    {
+      if (errno == EINVAL)
+        res = TEMP_FAILURE_RETRY (flock (fd, flags));
+      if (res < 0)
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+get_repo_lock (OstreeRepo          *repo,
+               int                  flags,
+               int                 *out_lock_fd,
+               GCancellable        *cancellable,
+               GError             **error)
+{
+  glnx_autofd int lock_fd = -1;
+
+  /* This re-implements a simpler (non-stacking) version of the ostree repo lock, as
+     the API for that is not yet available. When it is (see https://github.com/ostreedev/ostree/pull/2341)
+     this should be removed.
+     Note: This also doesn't respect the locking config options, it always locks and it always blocks.
+  */
+
+  lock_fd = TEMP_FAILURE_RETRY (openat (ostree_repo_get_dfd (repo), ".lock",
+                                        O_CREAT | O_RDWR | O_CLOEXEC, 0660));
+  if (lock_fd < 0)
+    return glnx_throw_errno_prefix (error,
+                                    "Opening lock file %s/.lock failed",
+                                    flatpak_file_get_path_cached (ostree_repo_get_path (repo)));
+
+  if (!do_repo_lock (lock_fd, flags))
+    return glnx_throw_errno_prefix (error, "Locking repo exclusively failed");
```

Comment:
Couldn't the lock have failed even if it was a shared lock not exclusive, in which case this error message would be misleading?